### PR TITLE
Track delivered projects

### PR DESCRIPTION
**Pull request by johandahlberg**
I'm sorry for dumping such a big PR, but there are so many interconnected things in this  that it's difficult to restrict it. This is work in progress on two issues. 1) Being able to deliver multiple flowcells for a project. 2) Tracking which projects/flowcells/etc have already been delivered.

At this stage I'm fairly confident that the staging and delivery functions for single runfolder and arbitrary directory (i.e. project) deliveries are working as expected. The "multiple flowcells for a project" needs some more testing, both manual an automatic before I feel confident it is ok.

I addition to adding new features here I've stared refactoring of the code in order to make the handlers and services less complex. This PR introduced a new service, the `DeliveryService` which functions as a hub for many of the other features, moving lots of functionality which was previously in the `StagingService` into it. Hopefully this will make more sense long term.

To summarize, at least these things still need to be carried out for this to be considered finished:
 - [x] Ensure that non of the asynchronous features have broken (i.e. having one staging or delivery on-going should not block the entire service)
 - [x] Test with the arteria-packs workflows ensuring that they are still compatible (and add the necessary parameters to them for forcing a re-delivery)
 - [x] Write integration tests for "multiple flowcell delivery"
 - [x] Run manual testing for "multiple flowcell delivery"
 - [x] Do a round and make sure all this refactoring has not introduced any dead code.

### CODE REVIEW

**Comment by johandahlberg** [comment]
This should be ready for review now.

There might be some issue with the concurrency, though I've not been able to solve it, and I've never seen it when running things manually (only in the automatic integration tests). However, at this point I'm not entierly sure whether the problems lies in the test code or the app code. My current suggestion is to solve that if it turns out to be a problem in reality later.

File: delivery/services/staging_service.py [hunk 1]
```diff
@@ -194,6 +203,61 @@ def stage_runfolder(self, runfolder_id, projects_to_stage=None, callback=None):
...
+        :param project_name: name of the project
+        :param projects: runfolders with the specified project on them
+        :return: the path to the dir created
+        """
+
+        project_dir = os.path.join(self.project_links_directory, project_name)
+        try:
+            FileSystemService.mkdir(project_dir)
+        except FileExistsError as e:
+            log.warning("Project dir: {} already exists".format(project_dir))
```
Comment (johanherman) [inline delivery/services/staging_service.py]: Should it continue if it exists?

File: delivery/handlers/staging_handlers.py [hunk 2]
```diff
@@ -153,6 +153,10 @@ def post(self, directory_name):
                 'content-type': "application/json",
             }
 
+            # Optionally send a project alias (when the name of the dir is something else
+            than the project name)
```
Comment (johanherman) [inline delivery/handlers/staging_handlers.py]: Add a missing `#` so that the nested comments doesn't confuse as much. :)

File: delivery/services/staging_service.py [hunk 3]
```diff
@@ -194,6 +203,61 @@ def stage_runfolder(self, runfolder_id, projects_to_stage=None, callback=None):
...
+        return FileSystemService.abspath(project_dir)
+
+    def stage_runfolders_for_project(self, project_id, delivery_type):
+        projects = list(self.runfolder_project_repo.get_project(project_name=project_id))
+
+        if len(projects) < 1:
+            raise ProjectNotFoundException("Could not find any Project folders"
+                                           " for project name: {}".format(project_id))
+
+        # TODO
```
Comment (johanherman) [inline delivery/services/staging_service.py]: I presume these will be defined somewhere, as I noticed you were referencing these in `StagingProjectRunfoldersHandler` as well.

File: delivery/models/delivery_modes.py [hunk 4]
```diff
@@ -0,0 +1,7 @@
+from enum import Enum
+
+
+class DeliveryMode(Enum):
```
Comment (johanherman) [inline delivery/models/delivery_modes.py]: Perhaps a simple explanation of each mode would be warranted.
